src/app/stores/action_handlers/document.ts

```diff
@@ -247,25 +244,101 @@ export default function(REG: ActionHandlerRegistry<AppStore, Actions.Action>) {
     this.emit(AppStore.EVENT_SELECTION);
   });
 
-  REG.add(Actions.ReplaceDataset, function(action) {
+  REG.add(Actions.UpdatePlotSegments, function (action) {
+    this.saveHistory();
+
+    this.updatePlotSegments();
+    this.solveConstraintsAndUpdateGraphics();
+    this.emit(AppStore.EVENT_DATASET);
+    this.emit(AppStore.EVENT_SELECTION);
+  });
+
+  REG.add(Actions.ReplaceDataset, function (action) {
     this.saveHistory();
 
     this.currentChartID = null;
     this.currentSelection = null;
     this.dataset = action.dataset;
+    this.originDataset = deepClone(this.dataset);
 
     this.chartManager = new Prototypes.ChartStateManager(
       this.chart,
       this.dataset
     );
     this.chartState = this.chartManager.chartState;
-
+    this.updatePlotSegments();
     this.solveConstraintsAndUpdateGraphics();
     this.emit(AppStore.EVENT_DATASET);
     this.emit(AppStore.EVENT_SELECTION);
   });
 
-  REG.add(Actions.Undo, function(action) {
+  REG.add(Actions.ConvertColumnDataType, function (action) {
+    this.saveHistory();
+
+    const table = this.dataset.tables.find(table => table.name === action.tableName);
+    if (!table) {
+      return;
+    }
+
+    const column = table.columns.find(column => column.name === action.column);
+
+    if (!column) {
+      return;
+    }
+
+    const applyConvertedValues = (table: Table, columnName: string, convertedValues: Array<(string | number | boolean)>) => {
+      table.rows.forEach((value: any, index: number) => {
+        table.rows[index][columnName] = convertedValues[index];
```

Comment:
Can you use `value` instead of `table.rows[index]`  here to avoid the extra lookup?

Reply:
Updated